### PR TITLE
feat: Workbook Export Plugin

### PR DESCRIPTION
**Pull request by hansjhoffman**
# Workbook Export Plugin

Initially designed to export all data from a Flatfile Workbook into a newly created Excel Workbook with tabs corresponding to each Flatfile Sheet. Currently only exports to `.xlsx` files but the underlying library has many more to offer. 

Includes:
- Debug mode for development since the plugin can fail in numerous ways making it hard to debug
- Cell comments generated from `FlatfileRecord` messages (currently limited to 104 columns -- self imposed limit)

---

[Watch video](https://www.loom.com/share/bd982df23ddf4f4a9dec12ac9fbc5311?sid=2654f01b-5092-40bd-a18f-820fc156f81c)

Screenshot 2023-07-25 at 2 16 36 PM

Screenshot 2023-07-25 at 2 16 46 PM

Screenshot 2023-07-25 at 1 06 31 PM

Screenshot 2023-07-19 at 1 24 24 PM

### Usage

```sh
curl --request POST \
  --url https://platform.flatfile.com/api/v1/workbooks \
  --header 'Authorization: Bearer [secret_key]' \
  --header 'Content-Type: application/json' \
  --data '{
  "name": "My First Workbook",
  "sheets": [
    {
      "name": "Contact",
      "slug": "contacts",
      "fields": [
        {
          "key": "firstName",
          "type": "string",
          "label": "First Name"
        },
        {
          "key": "lastName",
          "type": "string",
          "label": "Last Name"
        },
        {
          "key": "email",
          "type": "string",
          "label": "Email"
        }
      ]
    }
  ],
  "actions": [
    {
      "operation": "downloadExcelWorkbook",
      "mode": "foreground",
      "label": "Download Excel Workbook",
      "description": "Downloads Excel Workbook of Data",
      "primary": true
    }
  ]
}'
```

```ts
import { Client } from "@flatfile/listener";
import { xlsxSinkPlugin } from "@flatfile/plugin-xlsx-workbook-sink";

export default function flatfileEventListener(listener: Client) {
  listener.use(xlsxSinkPlugin({ debug: process.env.NODE_ENV === "dev" }));
}
```

```sh
NODE_ENV=dev yarn dev
```

### CODE REVIEW

File: plugins/xlsx-worbook-sink/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# @flatfile/plugin-xlsx-workbook-sink"
+
+asdf...
+
+`npm i @flatfile/plugin-xlsx-workbook-sink`
```
Comment (hansjhoffman) [inline plugins/xlsx-worbook-sink/README.md:5]: ❔ What do we want to call plugins to "export/egress" data from Flatfile. I just chose "sink" b/c I'm used to Kafka but I'm open to anything

File: plugins/xlsx-worbook-sink/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import { PluginOptions, run } from "./plugin";
+
+/**
+ * Excel Workbook export plugin for Flatfile.
+ *
+ * @param opts - plugin config options
+ */
+export const xlsxSinkPlugin = (opts: PluginOptions = {}) => {
+  return (listener: FlatfileListener) => {
+    listener.filter({ job: "workbook:downloadExcelWorkbook" }, () => {
```
Comment (hansjhoffman) [inline plugins/xlsx-worbook-sink/src/index.ts:12]: For now, job name will be hardcoded but we can add a config param later if necessary.

File: plugins/xlsx-worbook-sink/src/plugin.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+      return;
+    }
+  } catch (_fetchSheetsError: unknown) {
+    logError("Failed to fetch sheets for workbook id: " + workbookId);
+
+    return;
+  }
+};
+
+const currentEpoch = (): string => {
```
Comment (hansjhoffman) [inline plugins/xlsx-worbook-sink/src/plugin.ts:161]: If we don't like epoch as a unique identifier on files, we can change it something else.

File: plugins/xlsx-worbook-sink/src/plugin.spec.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+
```
Comment (hansjhoffman) [inline plugins/xlsx-worbook-sink/src/plugin.spec.ts:1]: No easy way to test this. I could use cypress or playwright to write a full e2e test if that's what we want.

**Review by ashleygmulligan2 — APPROVED** [review]

**Comment by ashleygmulligan2** [comment]
what is NODE_ENV=dev yarn dev for?

File: plugins/xlsx-worbook-sink/src/index.ts (hunk 2 above)
Comment (hansjhoffman) [reply plugins/xlsx-worbook-sink/src/index.ts]: It was designed to export ALL sheets in a Flatfile workbook into an excel-like workbook therefore, probably only on a workbook.

**Comment by ashleygmulligan2** [comment]
next pass: props for workbook or sheet action, and the name of the operation. Possible let this also work at Space level once Space Actions are a thing

**Comment by ashleygmulligan2** [comment]
Next pass: Think through a headless use case and see if this fits

**Comment by ashleygmulligan2** [comment]
Please update PR description with new naming

**Comment by carlbrugger** [comment]
@hansjhoffman the directory appears to have a typo: `plugins/worbook`.

**Comment by carlbrugger** [comment]
@hansjhoffman no, the actual directory name: 
Screenshot 2023-07-26 at 10 01 23

**Pull request merged**